numba_cuda/numba/cuda/tests/cudadrv/test_module_callbacks.py

```diff
@@ -0,0 +1,201 @@
+
+        self.assertEqual(len(setup_seen), 2)
+        self.assertEqual(len(teardown_seen), 2)
+
+    def test_two_kernels(self):
+        counter = 0
+        setup_seen = set()
+        teardown_seen = set()
+
+        def setup(handle):
+            nonlocal counter, setup_seen
+            counter += 1
+            setup_seen.add(handle.value)
+
+        def teardown(handle):
+            nonlocal counter, teardown_seen
+            counter -= 1
+            teardown_seen.add(handle.value)
+
+        lib = CUSource("", setup_callback=setup, teardown_callback=teardown)
+
+        @cuda.jit(link=[lib])
+        def kernel():
+            pass
+
+        @cuda.jit(link=[lib])
+        def kernel2():
+            pass
+
+        kernel[1, 1]()
+        self.assertEqual(counter, 1)
+        kernel2[1, 1]()
+        self.assertEqual(counter, 2)
+
+        wipe_all_modules_in_context()
+        del kernel
+        self.assertEqual(counter, 0)
+
+        self.assertEqual(len(setup_seen), 2)
+        self.assertEqual(len(teardown_seen), 2)
+
+
+class TestModuleCallbacksAPICompleteness(CUDATestCase):
+
+    def test_api(self):
+        def setup(handle):
+            pass
+
+        def teardown(handle):
+            pass
+
+        api_combo = [
+            (setup, teardown),
+            (setup, None),
+            (None, teardown),
+            (None, None)
+        ]
+
+        for setup, teardown in api_combo:
+            with self.subTest(setup=setup, teardown=teardown):
+                lib = CUSource(
+                    "", setup_callback=setup, teardown_callback=teardown)
+
+                @cuda.jit(link=[lib])
+                def kernel():
+                    pass
+
+                kernel[1, 1]()
+
+
+class TestModuleCallbacks(CUDATestCase):
+
+    def setUp(self):
+        super().setUp()
+
+        module = """
+__device__ int num = 0;
+extern "C"
+__device__ int get_num(int &retval) {
+    retval = num;
+    return 0;
+}
+"""
+
+        def set_forty_two(handle):
+            # Initialize 42 to global variable `num`
+            res, dptr, size = cuModuleGetGlobal(
+                handle.value, "num".encode()
+            )
+
+            arr = np.array([42], np.int32)
+            cuMemcpyHtoD(dptr, arr.ctypes.data, size)
+
+        self.lib = CUSource(
+            module, setup_callback=set_forty_two, teardown_callback=None)
+
+    def test_decldevice_arg(self):
+        get_num = cuda.declare_device("get_num", "int32()", link=[self.lib])
+
+        @cuda.jit
+        def kernel(arr):
+            arr[0] = get_num()
+
+        arr = np.zeros(1, np.int32)
+        kernel[1, 1](arr)
+        self.assertEqual(arr[0], 42)
+
+    def test_jitarg(self):
+        get_num = cuda.declare_device("get_num", "int32()")
+
+        @cuda.jit(link=[self.lib])
+        def kernel(arr):
+            arr[0] = get_num()
+
+        arr = np.zeros(1, np.int32)
+        kernel[1, 1](arr)
+        self.assertEqual(arr[0], 42)
```

Comment:
We should also be able to run the test file standalone:

```suggestion
        self.assertEqual(arr[0], 42)
        
if __name__ == '__main__':
    unittest.main()
```